service/raft/raft_group0.cc

```diff
@@ -424,11 +512,65 @@ future<> raft_group0::leave_group0() {
         co_return;
     }
 
-    if (!joined_group0()) {
-        // TODO: unimplemented upgrade case.
+    auto upgrade_state = (co_await _client.get_group0_upgrade_state()).second;
+    if (upgrade_state == group0_upgrade_state::recovery) {
+        group0_log.warn("leave_group0: in Raft RECOVERY mode, skipping.");
+        co_return;
+    }
+
+    if (!_feat.supports_raft_cluster_mgmt) {
+        // The Raft feature is not yet enabled.
+        //
+        // In that case we assume the cluster is in a partially-upgraded state (some nodes still use a version
+        // without the raft/group 0 code enabled) and skip the leaving group 0 step.
+        //
+        // In theory we could be wrong: the cluster may have just upgraded but the user managed to call decommission
+        // before we noticed it. Us leaving at this point could cause `upgrade_to_group0`
+        // on other nodes to get stuck because it requires contacting all peers which may include us.
+        //
+        // For this to happen, the following conditions would have to be true:
+        // 1. decommission is called immediately after upgrade, before we notice that the Raft feature is enabled
+        // 2. we don't notice it during decommission - including the long streaming phase - even until we reach
+        //    the `leave_group0` step which is almost at the end of the procedure
+        // 3. some node did notice that the feature is enabled and started `upgrade_to_group0`, using us
+        //    as one of the peers for running group 0 discovery, before we entered LEFT state
```

Comment:
Could this race be fixed by introducing another cluster feature which is marked supported when a node joins group0? Then each node would wait for this feature to be enabled globally before proceeding with synchronization. This ensures that when we start synchronizing, every node sees supports_raft_cluster_mgmt as enabled and will know to maintain group0 membership.

Reply:
Today I gave up on cluster features completely. The cluster features feature is a broken piece of crap. It's unreliable. A feature can get enabled even though not all nodes support it. I tried to fix this, but then it unmasked other problems with cluster features.

See: https://github.com/scylladb/scylladb/pull/11225
Which introduced a regression: https://github.com/scylladb/scylladb/issues/11316#issuecomment-1219381871

Now I'm considering adding another step at the beginning of the upgrade procedure, that will RPC all nodes before doing anything else in order to confirm that they really know about the upgrade procedure. So basically reimplementing what cluster features was supposed to provide.

Reply:
I think there is about 0 value in thinking about how to fix this race which will probably never happen - you'd have to decommission a node and go through streaming and all before the upgrade procedure finishes.

> introducing another cluster feature which is marked supported when a node joins group0

Please let's not try to implement the steps in terms of cluster features, given that:
1. they don't have the ability to rollback (which we need recovery)
2. they are a broken piece of crap anyway, and relying on them is a mistake.

Reply:
>  Then each node would wait for this feature to be enabled globally before proceeding with synchronization. This ensures that when we start synchronizing, every node sees supports_raft_cluster_mgmt as enabled and will know to maintain group0 membership.

The problem described in this scenario happens (or can happen?) before even joining group 0, the very first step of the upgrade procedure. The problem is that some node would start `upgrade_to_group0` and run discovery procedure (to discover group 0 peers) including us as seeds (so it would have to contact us to finish discovery) and it would get stuck because it wouldn't be able to contact us.

Reply:
It's very sad that cluster features are unreliable, we should fix that regardless rather than have each subsystem reinvent its own analog. I won't insist to do it now, but the abstraction itself makes sense, and makes algorithms easy to reason about due to monotonicity. Non-monotonic state changes leads to explosion of possible scenarios.

Recovery is possible with cluster features, you can restore database state from backup by downing all nodes and upping all nodes. You lose availability, but at least you know you're in a safe state. Hot recovery is better, but adds complexity and so is a risky thing, also to maintain. It may work now, but as the code evolves it may bit rot.

Reply:
> It may work now, but as the code evolves it may bit rot.

Fortunately, it's easy to test, because Raft will be disable-able with a flag even when it leaves experimental.
We'll test it in the new test.py cluster framework.